js/runner.ts

```diff
@@ -0,0 +1,184 @@
+      // is the case with JSON modules and potentially other future features.
+      const exports = module.factory!(...this._getFactoryArguments(module));
+      if (exports != null) {
+        module.exports = exports;
+      }
+    }
+  }
+
+  private _gatherDependencies(filename: Filename): void {
+    log("runner._gatherDependencies", filename);
+
+    if (this._modules.has(filename)) {
+      log("Module already exists:", filename);
+      return;
+    }
+
+    const module = new Module(filename, {});
+    this._modules.set(filename, module);
+
+    window.define = this._makeDefine(module);
+    globalEval(this._provider.getOutput(filename));
+    window.define = undefined;
+  }
+
+  // tslint:disable-next-line:no-any
+  private _getFactoryArguments(module: Module): any[] {
+    log("runner._getFactoryArguments", module.filename);
+    assert(module.deps != null, "Missing dependencies for module.");
+    return module.deps!.map(dep => {
+      if (dep === "require") {
+        return this._makeLocalRequire(module);
+      }
+      if (dep === "exports") {
+        return module.exports;
+      }
+      if (dep in Runner._builtins) {
+        return Runner._builtins[dep];
+      }
+      const depModule = this._modules.get(dep)!;
+      assert(dep != null, `Missing dependency "${dep}"`);
+      return depModule.exports;
+    });
+  }
+
+  private _makeDefine(module: Module): AmdDefine {
+    log("runner._makeDefine", module.filename);
+    return (deps: ModuleSpecifier[], factory: AmdFactory): void => {
+      module.factory = factory;
+      module.deps = deps.map(dep => {
+        if (dep === "require" || dep === "exports" || dep in Runner._builtins) {
+          return dep;
+        }
+        const depFilename = this._provider.getFilename(dep, module.filename);
+        if (!this._modules.get(depFilename)) {
+          this._gatherDependencies(depFilename);
+        }
+        return depFilename;
+      });
+      if (!this._runQueue.includes(module)) {
+        this._runQueue.push(module);
+      }
+    };
+  }
+
+  private _makeLocalRequire(module: Module): AmdRequire {
+    log("runner._makeLocalRequire", module.filename);
+    return (
+      deps: ModuleSpecifier[],
+      callback: AmdCallback,
+      errback: AmdErrback
+    ): void => {
+      log("runner._makeLocalRequire", deps);
+      assert(
+        deps.length === 1,
+        "Local require supports exactly one dependency."
+      );
+      const [moduleSpecifier] = deps;
+      try {
+        this.run(moduleSpecifier, module.filename);
+        const requiredFilename = this._provider.getFilename(
+          moduleSpecifier,
+          module.filename
+        );
+        const requiredModule = this._modules.get(requiredFilename)!;
+        assert(requiredModule != null);
+        callback(requiredModule.exports);
+      } catch (e) {
+        errback(e);
+      }
+    };
+  }
+
+  constructor(provider: Provider) {
+    this._provider = provider;
+  }
+
```

Comment:
Maybe we could add `incrementalRun(compiledCode: OutputCode, previousCode: OutputCode)`? It would be used to run REPL code and potentially to evaluate string of code passed when deno is run (`deno --evaluate Date.now()`)

Reply:
Maybe, but that would be something that should be done after this lands.  We should try to break things into incremental units (partly why breaking up the isolate changes into smaller changes).